components/org.wso2.carbon.identity.application.authenticator.biometric/src/main/java/org/wso2/carbon/identity/application/authenticator/biometric/notification/handler/impl/FirebasePushNotificationSenderImpl.java

```diff
@@ -0,0 +1,170 @@
+
+            JSONObject biometricNotificationInfo = new JSONObject();
+            biometricNotificationInfo.put(BODY, message);
+            biometricNotificationInfo.put(CONTENT_AVAILABLE, true);
+            biometricNotificationInfo.put(PRIORITY, HIGH);
+
+            JSONObject biometricNotificationData = new JSONObject();
+            biometricNotificationData.put(BODY, message);
+            biometricNotificationData.put(CHALLENGE, randomChallenge);
+            biometricNotificationData.put(SESSION_KEY, sessionDataKey);
+            //Reason for sending the click_action in the data payload is to
+            // specifically open a different activity in android app except the default main activity.
+            biometricNotificationData.put(CLICK_ACTION, DISPLAY_ANDROID_ACTIVITY);
+            biometricNotificationData.put(CONTENT_AVAILABLE, true);
+            biometricNotificationData.put(PRIORITY, HIGH);
+
+            JSONObject json = new JSONObject();
+            json.put(TO, deviceId.trim());
+            json.put(NOTIFICATION, biometricNotificationInfo);
+            json.put(DATA, biometricNotificationData);
+            json.put(CONTENT_AVAILABLE, true);
+            json.put(PRIORITY, HIGH);
+
+            if (log.isDebugEnabled()) {
+                log.debug("Firebase message payload: " + json.toString());
+            }
+
+            OutputStreamWriter wr = null;
+            try {
+                wr = new OutputStreamWriter(conn.getOutputStream(), StandardCharsets.UTF_8);
+                wr.write(json.toString());
+                wr.flush();
+            } finally {
+                if (wr != null) {
+                    wr.close();
+                }
+            }
+
+            int status = conn.getResponseCode();
+            if (status != HttpServletResponse.SC_OK) {
+                if (status == HttpServletResponse.SC_BAD_REQUEST) {
+                    log.error("Request parameters were invalid.");
+
+                } else if (status == HttpServletResponse.SC_UNAUTHORIZED) {
+                    log.error("Notification Response : Device Id : " + deviceId + " Error occurred.");
+
+                } else if (status == HttpServletResponse.SC_NOT_FOUND) {
+                    log.error("App instance was unregistered from FCM.Token used is no longer valid. ");
+
+                } else if (status == HttpServletResponse.SC_INTERNAL_SERVER_ERROR) {
+                    log.error(" An unknown internal error occurred.");
+
+                } else if (status == HttpServletResponse.SC_SERVICE_UNAVAILABLE) {
+                    log.error("The server is overloaded. DeviceId : " + deviceId);
+                } else {
+                    log.error("Some unknown error occurred when initiating the push notification. ");
+                }
+            }
+
+
+        } catch (MalformedURLException e) {
+            throw new AuthenticationFailedException("Invalid URL ", e);
```

Comment:
Let's make the error message more descriptive. ex: add the URL

Reply:
logged the valid url to be entered.